lib/TH/generic/THStorage.h

```diff
@@ -28,6 +29,8 @@ typedef struct THStorage
     char flag;
     THAllocator *allocator;
     void *allocatorContext;
+    int type;
```

Comment:
So the main reason for this is the (not so great) templating system in C. Our torch env is wrapped in different types of environments (Java/Scala/JNI), and the way these wrappers are written is without templating. They discover the type of the storage at runtime, and then act appropriately. 

It's pretty useful for us.

About the Cuda type, it was convenient to have them grouped in a same header, but fair point.